### PR TITLE
mesa: add vulkanLayers option and default it to device-select, overlay

### LINKED ISSUE #196086: mesa: build doesn't include device-select and overlay layers
### Describe the bug
Our mesa build is missing the device-select and overlay layers so that functionality doesn't work.

Add `"-D vulkan-layers=device-select,overlay"` to mesonFlags and 

```
      mv -t $drivers/lib $out/lib/libVkLayer*
      for js in $drivers/share/vulkan/implicit_layer.d/*.json; do
        substituteInPlace "$js" --replace '"libVkLayer_' '"'"$drivers/lib/libVkLayer_"
      done
```

to postInstall to get device-select working. This doesn't get the overlay working, probably needs a similar change.

Other distros eg arch do include these layers: https://github.com/archlinux/svntogit-packages/blob/a66847a882888b0c2e1f55085224ba8ac2fb3474/trunk/PKGBUILD#L56

The functionality these layers enable is described [here for device-select](https://docs.mesa3d.org/envvars.html#vulkan-mesa-device-select-layer-environment-variables).

### PR DESCRIPTION
**Pull request by LunNova**
Fixes #196086

###### Description of changes

Enables building the optional device-select and overlay vulkan layers, and fixes the paths so they work.

There's currently one more optional layer `intel-nullhw` but it seems much less useful.

Arch actually [does turn on intel-nullhw](https://github.com/archlinux/svntogit-packages/blob/a66847a882888b0c2e1f55085224ba8ac2fb3474/trunk/PKGBUILD#L56) by default so maybe we should include it too, haven't tracked down why they have it on.

Commands to test the layers:

```
# Should launch vkcube with overlay on
VK_INSTANCE_LAYERS=VK_LAYER_MESA_overlay VK_LAYER_MESA_OVERLAY_CONFIG=help vkcube

# Should show only one GPU in vulkaninfo using the device select layer
`MESA_VK_DEVICE_SELECT_FORCE_DEFAULT_DEVICE=1` vulkaninfo --summary

# Compare against
vulkaninfo --summary
```

Here's the overlay in action:

image

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @PedroHLC

### CODE REVIEW

File: pkgs/development/libraries/mesa/default.nix [hunk 1]
```diff
@@ -119,7 +120,8 @@ self = stdenv.mkDerivation {
     "-Dgallium-opencl=icd" # Enable the gallium OpenCL frontend
     "-Dclang-libdir=${llvmPackages.clang-unwrapped.lib}/lib"
   ] ++ optional enablePatentEncumberedCodecs
-    "-Dvideo-codecs=h264dec,h264enc,h265dec,h265enc,vc1dec";
+    "-Dvideo-codecs=h264dec,h264enc,h265dec,h265enc,vc1dec"
+  ++ lib.optionals (vulkanLayers != []) [ "-D vulkan-layers=${builtins.concatStringsSep "," vulkanLayers}" ];
```
Comment (PedroHLC) [inline pkgs/development/libraries/mesa/default.nix:124]: You could simplify it to use `optional`.

**Review by PedroHLC — APPROVED** [review]
Looks good. I can confirm it evaluates, builds and works.

We can update this part of the wiki after merging it: https://nixos.wiki/wiki/Mesa#Enabling_Vulkan_Layers

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (LunNova) [reply pkgs/development/libraries/mesa/default.nix]: Done

**Comment by PedroHLC** [comment]
@vcunat and/or @primeos, thoughts?

**Comment by LunNova** [comment]
FYI I found some upstream bugs in device-select but they mostly only apply to people on 3+ GPU systems which is rare, and I still think it's better to have this in since the majority of systems with 1 or 2 GPUs work fine. It's not any worse than not having it.

<https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/19101>

**Review by thiagokokada — APPROVED** [review]

**Pull request merged**